ada_feeding_perception/ada_feeding_perception/segment_all_items.py

```diff
@@ -0,0 +1,1279 @@
+
+        Parameters
+        ----------
+        image: The image to perform segmentation on.
+        seed_point: The seed point for EfficientSAM to segment from.
+        bbox: The bounding box prompt for EfficientSAM to segment.
+        prompt: The prompt to use for EfficientSAM. If 0, use the seed point prompt.
+                If 1, use the bounding box prompt.
+
+        Returns
+        -------
+        masks: The masks for each segmentation.
+        scores: The confidence scores for each segmentation.
+        """
+        self._node.get_logger().info("Segmenting image with EfficientSAM...")
+
+        # Convert image from BGR to RGB for Segment Anything
+        image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
+
+        # Convert the image to a tensor
+        image_tensor = transforms.ToTensor()(image).to(self.device)
+
+        # Convert input prompt (seed point or bounding box) to a tensor
+        if prompt == 0:
+            prompt_tensor = torch.tensor(np.array(seed_point).reshape((1, 1, 1, 2))).to(
+                device=self.device
+            )
+
+            # Define the labels for the input prompt
+            prompt_labels = torch.tensor([[[1]]]).to(device=self.device)
+        else:
+            prompt_tensor = torch.reshape(torch.tensor(bbox), [1, 1, 2, 2]).to(
+                self.device
+            )
+
+            # Define the labels for the input prompt
+            prompt_labels = torch.reshape(torch.tensor([2, 3]), [1, 1, 2]).to(
+                self.device
+            )
+
+        # Run EfficientSAM on the image using the input prompt
+        predicted_logits, predicted_iou = self.efficient_sam(
+            image_tensor[None, ...],
+            prompt_tensor,
+            prompt_labels,
+        )
+        sorted_ids = torch.argsort(predicted_iou, dim=-1, descending=True)
+        predicted_iou = torch.take_along_dim(predicted_iou, sorted_ids, dim=2)
+        predicted_logits = torch.take_along_dim(
+            predicted_logits, sorted_ids[..., None, None], dim=2
+        )
+        masks = torch.ge(predicted_logits[0, 0, :, :, :], 0).cpu().detach().numpy()
+        scores = predicted_iou[0, 0, :].cpu().detach().numpy()
+
+        return masks, scores
+
+    def run_grounding_dino(
+        self,
+        image: npt.NDArray,
+        caption: str,
+        box_threshold: float,
+        text_threshold: float,
+    ):
+        """
+        Run GroundingDINO on the image.
+
+        Parameters
+        ----------
+        image: The CV2 above the plate image to retrieve semantically labeled bounding
+               boxes from.
+        caption: The caption to use as text input for GroundingDINO.
+        box_threshold: The threshold for the bounding box.
+        text_threshold: The threshold for the text.
+
+        Returns
+        -------
+        bbox_predictions: A dictionary containing the bounding boxes for each food item label
+                        detected from the image.
```

Comment:
Add here how bounding boxes are represented. I'm guessing it is `(x, y, w, h)` where `x,y` is the top-left corner?